lib/valueflow.cpp

```diff
@@ -2468,67 +2468,160 @@ static bool valueFlowForward(Token * const               startToken,
     return true;
 }
 
+static const Variable * getLifetimeVariable(const Token * tok, ErrorPath& errorPath)
+{
+    const Variable * var = tok->variable();
+    if (!var)
+        return nullptr;
+    if (var->isReference() || var->isRValueReference()) {
+        for (const ValueFlow::Value& v:tok->values()) {
+            if (!v.isLifetimeValue())
+                continue;
+            if (v.tokvalue == tok)
+                continue;
+            errorPath.insert(errorPath.end(), v.errorPath.begin(), v.errorPath.end());
+            const Variable * var2 = getLifetimeVariable(v.tokvalue, errorPath);
+            if (var2)
+                return var2;
+        }
+        return nullptr;
+    }
+    return var;
+}
+
 static bool isNotLifetimeValue(const ValueFlow::Value& val)
 {
     return !val.isLifetimeValue();
 }
 
+static void valueFlowLifetimeFunction(Token * tok, TokenList *tokenlist, ErrorLogger *errorLogger, const Settings *settings);
+
 static void valueFlowForwardLifetime(Token * tok, TokenList *tokenlist, ErrorLogger *errorLogger, const Settings *settings)
 {
-    const Token * assignTok = tok->astParent();
-    // Assignment
-    if (!assignTok || (assignTok->str() != "=") || assignTok->astParent())
-        return;
-
-    // Lhs should be a variable
-    if (!assignTok->astOperand1() || !assignTok->astOperand1()->varId())
-        return;
-    const Variable *var = assignTok->astOperand1()->variable();
-    if (!var || (!var->isLocal() && !var->isGlobal() && !var->isArgument()))
+    const Token * parent = tok->astParent();
+    while(parent && parent->isArithmeticalOp())
```

Comment:
coding style.. can you please try to get astyle working?

Reply:
I can format it with astyle 3.2, but i dont know how to set it to only format the lines I've changed. It seems astyle will always format the entire file, it doesn't have a `-lines` option to select only certain lines to be formatted. Do you have a script that can format only my changes?

Alternatively I could format it with clang-format. The problem with clang format is that it doesn't have an option to distinguish between inline and non-inline functions when attaching the brace.

Reply:
https://github.com/danmar/cppcheck/blob/master/runastyle.bat or https://github.com/danmar/cppcheck/blob/master/runastyle should be used to format the source tree.

They hardcode the astyle version to be used.

Reply:
Yea I dont have that astyle version. I can disable the version check, but that script formats every file. I just want to format the lines of code that I changed. Is there a script for that?

Reply:
> Using clang-format is discouraged, I assume it is highly incompatible.

Yep but it much easier to use than astyle.

> Why don't you just compile astyle locally.

That wont work on my mac, and the `runastyle` script won't run the built version either.

> I personally don't want to run make install because that is not "proper" way to install stuff. 

Manually copying binaries around is definitely not the proper way to "install".

> I always use sudo apt-get install.. to install system software..

I dont see how `sudo apt-get install` is related to `make install`. One is installing things into the system, the other is not. 

I would like to install astyle with `cget install https://sourceforge.net/projects/astyle/files/astyle/astyle%203.0.1/astyle_3.0.1_linux.tar.gz/download` , but that doesn't work. I can do that with the latest version off of trunk with `cget install https://sourceforge.net/code-snapshots/svn/a/as/astyle/code/astyle-code-r655-trunk-AStyle.zip`.

Reply:
ok thanks now I understand your problem better.

Reply:
I guess you have seen it. but they have separate source code archives for linux and mac. does it work better with:   

https://sourceforge.net/projects/astyle/files/astyle/astyle%203.0.1/astyle_3.0.1_macos.tar.gz/download

and I assume you have seen this:

 http://astyle.sourceforge.net/install.html#_MacOS_Version

if it just don't work for you then I guess a temporary hack for you is to use a sed script. If you remove trailing spaces and replace "if|while|for(" with "if|while|for (" that might fix some diffs.